packages/flutter_tools/lib/src/android/gradle.dart

```diff
@@ -200,456 +191,312 @@ void createSettingsAarGradle(Directory androidDirectory) {
+          // Don't pipe.
+          return null;
+        }
+        if (detectedGradleError != null) {
+          // Pipe stdout/sterr from Gradle.
+          return line;
+        }
+        for (final GradleHandledError gradleError in localGradleErrors) {
+          if (gradleError.test(line)) {
+            detectedGradleErrorLine = line;
+            detectedGradleError = gradleError;
+            // The first error match wins.
+            break;
+          }
+        }
+        // Pipe stdout/sterr from Gradle.
+        return line;
+      },
+    );
+  } finally {
+    status.stop();
   }
 
-  final File localProperties = project.android.localPropertiesFile;
-  bool changed = false;
+  flutterUsage.sendTiming('build', 'gradle', sw.elapsed);
 
-  SettingsFile settings;
-  if (localProperties.existsSync()) {
-    settings = SettingsFile.parseFromFile(localProperties);
-  } else {
-    settings = SettingsFile();
-    changed = true;
-  }
+  if (exitCode != 0) {
+    if (detectedGradleError == null) {
+      BuildEvent('gradle--unkown-failure').send();
+      throwToolExit(
+        'Gradle task $assembleTask failed with exit code $exitCode',
+        exitCode: exitCode,
+      );
+    } else {
+      final GradleBuildStatus status = await detectedGradleError.handler(
+        line: detectedGradleErrorLine,
+        project: project,
+        usesAndroidX: usesAndroidX,
+        shouldBuildPluginAsAar: shouldBuildPluginAsAar,
+      );
 
-  void changeIfNecessary(String key, String value) {
-    if (settings.values[key] != value) {
-      if (value == null) {
-        settings.values.remove(key);
-      } else {
-        settings.values[key] = value;
+      if (retries >= 1) {
+        final String successEventLabel = 'gradle--${detectedGradleError.eventLabel}-success';
+        switch (status) {
+          case GradleBuildStatus.retry:
+            await buildGradleApp(
+              project: project,
+              androidBuildInfo: androidBuildInfo,
+              target: target,
+              isBuildingBundle: isBuildingBundle,
+              localGradleErrors: localGradleErrors,
+              shouldBuildPluginAsAar: shouldBuildPluginAsAar,
+              retries: retries - 1,
+            );
+            BuildEvent(successEventLabel).send();
+            return;
+          case GradleBuildStatus.retryWithAarPlugins:
+            await buildGradleApp(
+              project: project,
+              androidBuildInfo: androidBuildInfo,
+              target: target,
+              isBuildingBundle: isBuildingBundle,
+              localGradleErrors: localGradleErrors,
+              shouldBuildPluginAsAar: true,
+              retries: retries - 1,
+            );
+            BuildEvent(successEventLabel).send();
+            return;
+          case GradleBuildStatus.exit:
+            // noop.
+        }
       }
-      changed = true;
+      BuildEvent('gradle--${detectedGradleError.eventLabel}-failure').send();
```

Comment:
You could include the exit code in one of the other fields of the event if that would be useful.